src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java

```diff
@@ -1204,11 +1238,15 @@ public Boolean visitClassType(ClassType t, Type s) {
                 if (sup == null) return false;
                 // If t is an intersection, sup might not be a class type
                 if (!sup.hasTag(CLASS)) return isSubtypeNoCapture(sup, s);
-                return sup.tsym == s.tsym
+                boolean result = sup.tsym == s.tsym
                      // Check type variable containment
                     && (!s.isParameterized() || containsTypeRecursive(s, sup))
                     && isSubtypeNoCapture(sup.getEnclosingType(),
                                           s.getEnclosingType());
+                if (result && warnStack.head.pos() != null && s.hasNarrowerNullabilityThan(t)) {
```

Comment:
I have memories of the previous implementation where we discussed maybe keeping subtyping routines as is, but adding some other check to validate the "polarity" of the type arguments. Wouldn't something similar work here too?